### PR TITLE
perf: perf get object workflow

### PR DESCRIPTION
**Pull request by will-2012**
### Description

Perf get-object workflow.

### Rationale

The split function is transferred from the downloader to the gater to reduce the memory usage of the downloader.

### Example

N/A.

### Changes

Notable changes: 
* Gater/Downloader: get-object workflow.
* Downloader: fix double push challenge queue bug.
* chore logs and error codes.

### CODE REVIEW

File: modular/gater/object_handler.go [hunk 1]
```diff
@@ -198,33 +199,45 @@ func (g *GateModular) getObjectHandler(w http.ResponseWriter, r *http.Request) {
...
-	w.Write(data)
-	log.CtxDebugw(reqCtx.Context(), "succeed to download object")
+	for idx, pInfo := range pieceInfos {
+		enableCheck := false
+		if idx == 0 { // only check in first piece
+			enableCheck = true
+		}
+		pieceTask := &gfsptask.GfSpDownloadPieceTask{}
+		pieceTask.InitDownloadPieceTask(objectInfo, bucketInfo, params, g.baseApp.TaskPriority(task),
+			enableCheck, reqCtx.Account(), uint64(highOffset-lowOffset+1), pInfo.SegmentPieceKey, pInfo.Offset,
```
Comment (joeylichang) [inline modular/gater/object_handler.go:232]: highOffset and lowOffset comes from object payload data size.
but get piece size should come from pieceInfo?

**Comment by joeylichang** [comment]
The add DonwloadPieceTask has done a lot of work. After the gater finishes compute the  pieceinfos, the range parameters of the download object can come from pieceseinfos each time(for range to call DonwloadObjectt rpc method), which can reuse a lot of previous logic and reduce the debugging workload.

File: modular/gater/object_handler.go (hunk 1 above)
Comment (will-2012) [reply modular/gater/object_handler.go:232]: Yes, `hightOffset/lowOffset` is the object‘s property used to check quota. And `pInfo.Offset/pInfo.Length` is the piece's property used to reserve memory resources.

**Comment by will-2012** [comment]
> The add DonwloadPieceTask has done a lot of work. After the gater finishes compute the pieceinfos, the range parameters of the download object can come from pieceseinfos each time(for range to call DonwloadObjectt rpc method), which can reuse a lot of previous logic and reduce the debugging workload.

The download object logic in the downloader module is a bit redundant for downloading pieces, and the direct multiplexing is a bit complicated. A separate interface is simpler.

**Review by joeylichang — APPROVED** [review]

File: base/gfspapp/download_server.go [hunk 2]
```diff
@@ -18,19 +18,17 @@ var (
 
 var _ gfspserver.GfSpDownloadServiceServer = &GfSpBaseApp{}
 
-func (g *GfSpBaseApp) GfSpDownloadObject(
-	ctx context.Context,
-	req *gfspserver.GfSpDownloadObjectRequest) (
+func (g *GfSpBaseApp) GfSpDownloadObject(ctx context.Context, req *gfspserver.GfSpDownloadObjectRequest) (
 	*gfspserver.GfSpDownloadObjectResponse, error) {
-	task := req.GetDownLoadTask()
+	task := req.GetDownloadObjectTask()
```
Comment (sysvm) [inline base/gfspapp/download_server.go:23]: the task var name is the same as imported package name, maybe use a new name.

File: base/gfspapp/download_server.go [hunk 3]
```diff
@@ -43,17 +41,15 @@ func (g *GfSpBaseApp) GfSpDownloadObject(
 		Data: data}, nil
 }
 
-func (g *GfSpBaseApp) OnDownloadObjectTask(
-	ctx context.Context,
-	task task.DownloadObjectTask) (
+func (g *GfSpBaseApp) OnDownloadObjectTask(ctx context.Context, task task.DownloadObjectTask) (
```
Comment (sysvm) [inline base/gfspapp/download_server.go:44]: ditto

File: base/gfspapp/download_server.go [hunk 4]
```diff
@@ -66,13 +62,55 @@ func (g *GfSpBaseApp) OnDownloadObjectTask(
 	return data, nil
 }
 
-func (g *GfSpBaseApp) GfSpGetChallengeInfo(
-	ctx context.Context,
-	req *gfspserver.GfSpGetChallengeInfoRequest) (
+func (g *GfSpBaseApp) GfSpDownloadPiece(ctx context.Context, req *gfspserver.GfSpDownloadPieceRequest) (
+	*gfspserver.GfSpDownloadPieceResponse, error) {
+	task := req.GetDownloadPieceTask()
```
Comment (sysvm) [inline base/gfspapp/download_server.go:67]: ditto

File: base/gfspapp/download_server.go [hunk 5]
```diff
@@ -66,13 +62,55 @@ func (g *GfSpBaseApp) OnDownloadObjectTask(
...
+	metrics.DownloadPieceSizeHistogram.WithLabelValues(
+		g.downloader.Name()).Observe(float64(task.GetSize()))
+	data, err := g.OnDownloadPieceTask(ctx, task)
+	log.CtxDebugw(ctx, "finished to download piece", "len", len(data), "error", err)
+	return &gfspserver.GfSpDownloadPieceResponse{
+		Err:  gfsperrors.MakeGfSpError(err),
+		Data: data}, nil
+}
+
+func (g *GfSpBaseApp) OnDownloadPieceTask(ctx context.Context, task task.DownloadPieceTask) (
```
Comment (sysvm) [inline base/gfspapp/download_server.go:88]: ditto. I found that task var name is too many, may change the imported package name.

File: base/gfspclient/downloader.go [hunk 6]
```diff
@@ -36,6 +36,31 @@ func (s *GfSpClient) GetObject(
 	return resp.GetData(), nil
 }
 
+func (s *GfSpClient) GetPiece(
+	ctx context.Context,
```
Comment (sysvm) [inline base/gfspclient/downloader.go:40]: maybe move these parameters to one line.

File: base/types/gfsptask/download.go [hunk 7]
```diff
@@ -169,6 +167,162 @@ func (m *GfSpDownloadObjectTask) SetBucketInfo(bucket *storagetypes.BucketInfo)
...
+	params *storagetypes.Params,
+	priority coretask.TPriority,
+	enableCheck bool,
+	userAddress string,
+	totalSize uint64,
+	pieceKey string,
+	pieceOffset uint64,
+	pieceLength uint64,
+	timeout int64,
+	maxRetry int64) {
```
Comment (sysvm) [inline base/types/gfsptask/download.go:182]: ditto

File: core/task/task.go [hunk 8]
```diff
@@ -404,6 +404,47 @@ type DownloadObjectTask interface {
...
+		object *storagetypes.ObjectInfo,
+		bucket *storagetypes.BucketInfo,
+		params *storagetypes.Params,
+		priority TPriority,
+		isFirstPiece bool,
+		userAddress string,
+		totalSize uint64,
+		pieceKey string,
+		pieceOffset uint64,
+		pieceLength uint64,
```
Comment (sysvm) [inline core/task/task.go:421]: ditto

File: modular/downloader/download_task.go [hunk 9]
```diff
@@ -171,15 +167,91 @@ func (d *DownloadModular) PostDownloadObject(
 	task task.DownloadObjectTask) {
 }
 
+func (d *DownloadModular) PreDownloadPiece(
+	ctx context.Context,
+	task task.DownloadPieceTask) error {
```
Comment (sysvm) [inline modular/downloader/download_task.go:172]: ditto

File: modular/downloader/download_task.go [hunk 10]
```diff
@@ -171,15 +167,91 @@ func (d *DownloadModular) PostDownloadObject(
...
+		}
+	}
+	// report the task to the manager for monitor the download piece task
+	d.baseApp.GfSpClient().ReportTask(ctx, task)
+	return nil
+}
+
+func (d *DownloadModular) HandleDownloadPieceTask(
+	ctx context.Context,
+	task task.DownloadPieceTask) ([]byte, error) {
```
Comment (sysvm) [inline modular/downloader/download_task.go:215]: ditto

File: modular/downloader/download_task.go [hunk 11]
```diff
@@ -171,15 +167,91 @@ func (d *DownloadModular) PostDownloadObject(
...
+	if pieceData, err = d.baseApp.PieceStore().GetPiece(ctx, task.GetPieceKey(),
+		int64(task.GetPieceOffset()), int64(task.GetPieceLength())); err != nil {
+		log.CtxErrorw(ctx, "failed to get piece data from piece store", "task_info", task.Info(), "error", err)
+		return nil, ErrPieceStore
+	}
+	return pieceData, nil
+}
+
+func (d *DownloadModular) PostDownloadPiece(
+	ctx context.Context,
```
Comment (sysvm) [inline modular/downloader/download_task.go:242]: ditto

File: base/gfspapp/download_server.go (hunk 2 above)
Comment (will-2012) [reply base/gfspapp/download_server.go]: Updated.

File: base/gfspapp/download_server.go (hunk 3 above)
Comment (will-2012) [reply base/gfspapp/download_server.go]: Updated.

File: base/gfspapp/download_server.go (hunk 4 above)
Comment (will-2012) [reply base/gfspapp/download_server.go]: Updated.

File: base/gfspapp/download_server.go (hunk 5 above)
Comment (will-2012) [reply base/gfspapp/download_server.go]: Updated.

File: base/gfspclient/downloader.go (hunk 6 above)
Comment (will-2012) [reply base/gfspclient/downloader.go]: Updated.

File: base/types/gfsptask/download.go (hunk 7 above)
Comment (will-2012) [reply base/types/gfsptask/download.go:182]: Updated.

File: core/task/task.go (hunk 8 above)
Comment (will-2012) [reply core/task/task.go]: Updated.

File: modular/downloader/download_task.go (hunk 9 above)
Comment (will-2012) [reply modular/downloader/download_task.go]: Updated.

File: modular/downloader/download_task.go (hunk 10 above)
Comment (will-2012) [reply modular/downloader/download_task.go]: Updated.

File: modular/downloader/download_task.go (hunk 11 above)
Comment (will-2012) [reply modular/downloader/download_task.go]: Updated.

**Review by joeylichang — APPROVED** [review]

**Review by sysvm — APPROVED** [review]
LGTM

**Pull request merged**